docs/man/mmcli.1
.\" mmcli(1) manual page
.\"
.\" Copyright (C) 2012 Martyn Russell
.\"
.\" Comment out '.nr' or set to 0 to eliminate WIDTH fiddlin' !
.nr half_xtra 4

.TH mmcli 1 "October 2012" GNU "User Commands"

.SH NAME
mmcli \- Control and monitor the ModemManager

.SH SYNOPSIS
\fBmmcli\fR [\fIOPTION\fR...]

.SH DESCRIPTION
ModemManager is a DBus-powered Linux daemon which provides a unified
high level API for communicating with (mobile broadband) modems. It
acts as a standard RIL (Radio Interface Layer) and may be used by
different connection managers, like NetworkManager. Thanks to the
built-in plugin architecture, ModemManager talks to very different
kinds of modems with very different kinds of ports. In addition to the
standard AT serial ports, Qualcomm-based QCDM and QMI ports are also
supported.

.SH HELP OPTIONS
.TP
.B \-h, \-\-help
Show summary of options by group.
.TP
.B \-\-help\-all
Show all groups and options.
.TP
.B \-\-help\-manager
Show manager specific options.
.TP
.B \-\-help\-common
Show common options. These are used for defining the device an option
operates on. For example, modems, bearers, SIMs, SMS', etc.
.TP
.B \-\-help\-modem
Show modem specific options.
.TP
.B \-\-help\-3gpp
Show 3GPP specific options.
.TP
.B \-\-help\-cdma
Show CDMA specific options.
.TP
.B \-\-help\-simple
Show simple options. These are useful for getting connected or
disconnected and understanding the state of things as fast as possible
without worrying so much about the details.
.TP
.B \-\-help\-location
Show location or positioning specific options.
.TP
.B \-\-help\-messaging
Show messaging specific options. See also \fB\-\-help\-sms\fR which
is related.
.TP
.B \-\-help\-cell-broadcast
Show cell broadcast specific options.
is related.
.TP
.B \-\-help\-time
Show time specific options.
.TP
.B \-\-help\-firmware
Show firmware specific options.
.TP
.B \-\-help\-oma
Show OMA specific options.
.TP
.B \-\-help\-sim
Show SIM card specific options.
.TP
.B \-\-help\-bearer
Show bearer specific options.
.TP
.B \-\-help\-sms
Show SMS specific options. See also \fB\-\-help\-messaging\fR which
is related.

.SH MANAGER OPTIONS
.TP
.B \-B, \-\-get\-daemon\-version
Retrieve the version of the currently running ModemManager daemon.
.TP
.B \-G, \-\-set\-logging=[ERR|WARN|INFO|DEBUG]
Set the logging level in ModemManager daemon. For debugging information you can
supply \fBDEBUG\fR. Each value above \fBDEBUG\fR provides less detail. In most
cases \fBERR\fR (for displaying errors) are the important messages.

The default mode is \fBERR\fR.
.TP
.B \-L, \-\-list\-modems
List available modems.
.TP
.B \-M, \-\-monitor\-modems
List available modems and monitor modems added or removed.
.TP
.B \-S, \-\-scan-modems
Scan for any potential new modems. This is only useful when expecting pure
RS232 modems, as they are not notified automatically by the kernel.
.TP
.B \-I, \-\-inhibit\-device=[UID]
Inhibit the specific device from being used by ModemManager. The \fBUID\fR
that should be given is the value of the \fBDevice\fR property exposed by
a given modem (i.e. equal to the \fBID_MM_PHYSDEV_UID\fR if one set, or
otherwise equal to the full device sysfs path).

This command will not exit right away, as that would implicitly remove the
inhibition. The user must make sure to stop the mmcli process hitting Ctrl+C
in order to un-inhibit the device.

When a device is inhibited via this method, ModemManager will disable the modem
(therefore stopping any ongoing connection) and will no longer use it until it
is uninhibited.
.TP
.B \-\-report\-kernel\-event=['KEY1=VALUE1,KEY2=VALUE2,...']
Manually report kernel events, instead of relying on udev (e.g. if the daemon
is running with \fB\-\-no\-auto\-scan\fR or if the system was built without udev
support).

The supported \fBKEY\fRs are:
.RS 9
.TP
\fB'action'\fR
Action to report, one of 'add' or 'remove'. Required.
.TP
\fB'subsystem'\fR
Subsystem of the specific port being reported, e.g. 'tty' (for serial ports),
\&'net' (for network interfaces), or 'usbmisc' (for cdc-wdm ports).
.TP
\fB'name'\fR
Name of the port being reported, e.g. 'ttyACM0', 'wwan0' or 'cdc-wdm0'.
.TP
\fB'uid'\fR
The specific UID of the device, equivalent to the \fBID_MM_PHYSDEV_UID\fR udev
tag. All ports reported with the same 'UID' value will be considered part of the
same device, which may be useful for e.g. modems with multiple platform TTYs.
.RE

.TP
.B \-\-report\-kernel\-event\-auto\-scan
When built with udev support but the daemon is running with
\fB\-\-no\-auto\-scan\fR, this method may be used to automatically report kernel
events based on udev.

This command will not exit right away. The user must make sure to stop the mmcli
process hitting Ctrl+C in order to stopping monitoring for new events.

.SH COMMON OPTIONS
All options below take a \fBPATH\fR or \fBINDEX\fR argument. If no action is
provided, the default information about the modem, bearer, etc. is
shown instead.

The \fBPATH\fR and \fBINDEX\fR are created automatically when the
modem is plugged in. They can be found using \fBmmcli \-L\fR. This
produces something like (for modems only):

.RS 7
Found 1 modems:
  /org/freedesktop/ModemManager1/Modem/\fB4\fR
.RE

In this case, the \fBINDEX\fR is \fB4\fR and the \fBPATH\fR is the entire string above.

However, for the bearers, SIMs and SMS cases, the \fBPATH\fR is
slightly different. The \fIModem\fR is replaced with the object name
in use, like \fIBearer\fR. For example:

.RS 7
/org/freedesktop/ModemManager1/\fIBearer\fR/4
.RE

.TP
.B \-m, \-\-modem=[PATH|INDEX]
Specify a modem.
.TP
.B \-b, \-\-bearer=[PATH|INDEX]
Specify a bearer.
.TP
.B \-i, \-\-sim=[PATH|INDEX]
Specify a SIM card.
.TP
.B \-s, \-\-sms=[PATH|INDEX]
Specify an SMS.

.SH MODEM OPTIONS
All of the modem options below make use of the \fB\-\-modem\fR or
\fB\-m\fR switch to specify the modem to act on.

Some operations require a \fBMODE\fR. \fBMODE\fR can be any
combination of the modes actually supported by the modem. In the
perfect case, the following are possible:

.Bd -literal -compact
  \fB'2G'\fR  - 2G technologies, e.g. EDGE, CDMA1x
  \fB'3G'\fR  - 3G technologies, e.g. HSPA, EV-DO
  \fB'4G'\fR  - 4G technologies, e.g. LTE
  \fB'ANY'\fR - for all supported modes.
.Ed
.TP
.B \-w, \-\-monitor\-state
Monitor the state of a given modem.
.TP
.B \-e, \-\-enable
Enable a given modem.

This powers the antenna, starts the automatic registration process and
in general prepares the modem to be connected.
.TP
.B \-d, \-\-disable
Disable a given modem.

This disconnects the existing connection(s) for the modem and puts it
into a low power mode.
.TP
.B \-r, \-\-reset
Resets the modem to the settings it had when it was power cycled.
.TP
.B \-\-factory\-reset=CODE
Resets the modem to its original factory default settings.

The \fBCODE\fR provided is vendor specific. Without the correct vendor
code, it's unlikely this operation will succeed. This is not a common
user action.
.TP
.B \-\-command=COMMAND
Send an AT \fBCOMMAND\fR to the given modem. For example,
\fBCOMMAND\fR could be 'AT+GMM' to probe for phone model information. This
operation is only available when ModemManager is run in debug mode.
.TP
.B \-\-create\-bearer=['KEY1=VALUE1,KEY2=VALUE2,...']
Create a new packet data bearer for a given modem. The \fBKEY\fRs and
some \fBVALUE\fRs are listed below:
.RS 9
.TP
\fB'apn'\fR
Access Point Name. Required in 3GPP.
.TP
\fB'ip-type'\fR
Addressing type. Given as a MMBearerIpFamily value (e.g. 'ipv4', 'ipv6', 'ipv4v6'). Optional in 3GPP and CDMA.
.TP
\fB'allowed-auth'\fR
Authentication method to use. Given as a MMBearerAllowedAuth value (e.g. 'none|pap|chap|mschap|mschapv2|eap'). Optional in 3GPP.
.TP
\fB'user'\fR
User name (if any) required by the network. Optional in 3GPP.
.TP
\fB'password'\fR
Password (if any) required by the network. Optional in 3GPP.
.TP
\fB'allow-roaming'\fR
Flag to tell whether connection is allowed during roaming, given as a boolean value (i.e 'yes' or 'no'). Optional in 3GPP.
.TP
\fB'rm-protocol'\fR
Protocol of the Rm interface, given as a MMModemCdmaRmProtocol value (e.g. 'async', 'packet-relay', 'packet-network-ppp', 'packet-network-slip', 'stu-iii'). Optional in CDMA.
.TP
\fB'number'\fR
Telephone number to dial. Required in POTS.
.RE
.TP
.B \-\-delete\-bearer=[PATH|INDEX]
Delete bearer from a given modem.
.TP
.B \-\-set\-allowed\-modes=[MODE1|MODE2|...]
Set allowed modes for a given modem. For possible modes, see the
beginning of this section.
.TP
.B \-\-set\-preferred\-mode=MODE
Set the preferred \fBMODE\fR for the given modem. The \fBMODE\fR
\fIMUST\fR be one of the allowed modes as set with the
\fB\-\-set\-allowed\-modes\fR option. Possible \fBMODE\fR arguments
are detailed at the beginning of this section.
.TP
.B \-\-set\-current\-bands=[BAND1|BAND2|...]
Set bands to be used for a given modem. These are frequency ranges
the modem should use. There are quite a number of supported bands and
listing them all here would be quite extensive. For details, see the
MMModemBand documentation.

An example would be: 'egsm|dcs|pcs|g850' to select all the GSM
frequency bands.
.TP
.B \-\-set\-primary\-sim\-slot=[SLOT]
Request to switch the primary SIM slot.

The given \fBSLOT\fR must be a valid slot number in the [1,N] range, where
N is the amount of SIM slots available in the system.
.TP
.B \-\-inhibit
Inhibit the specific modem from being used by ModemManager. This method
is completely equivalent to \fB\-\-inhibit\-device\fR, with the only
difference being that in this case, the modem must be managed by the daemon
at the time the inhibition is requested.

This command will not exit right away, as that would implicitly remove the
inhibition. The user must make sure to stop the mmcli process hitting Ctrl+C
in order to un-inhibit the device.

When a device is inhibited via this method, ModemManager will disable the modem
(therefore stopping any ongoing connection) and will no longer use it until it
is uninhibited.

.SH 3GPP OPTIONS
The 3rd Generation Partnership Project (3GPP) is a collaboration
between groups of telecommunications associations. These options
pertain to devices which support 3GPP.

Included are options to control USSD (Unstructured Supplementary
Service Data) sessions.

All of the 3GPP options below make use of the \fB\-\-modem\fR or
\fB\-m\fR switch to specify the modem to act on.
.TP
.B \-\-3gpp\-scan
Scan for available 3GPP networks.
.TP
.B \-\-3gpp\-register\-home
Request a given modem to register in its home network.

This registers with the default network(s) specified by the modem,
.TP
.B \-\-3gpp\-register\-in\-operator=MCCMNC
Request a given modem to register on the network of the given
\fBMCCMNC\fR (Mobile Country Code, Mobile Network Code) based
operator. This code is used for GSM/LTE, CDMA, iDEN, TETRA and UMTS
public land mobile networks and some satellite mobile networks. The
ITU-T Recommendation E.212 defines mobile country codes.
.TP
.B \-\-3gpp\-ussd\-status
Request the status of \fIANY\fR ongoing USSD session.
.TP
.B \-\-3gpp\-ussd\-initiate=COMMAND
Request the given modem to initiate a USSD session with \fBCOMMAND\fR.

For example, \fBCOMMAND\fR could be '*101#' to give your current
pre-pay balance.
.TP
.B \-\-3gpp\-ussd\-respond=RESPONSE
When initiating an USSD session, a \fRRESPONSE\fR may be needed by a
network-originated request. This option allows for that.
.TP
.B \-\-3gpp\-ussd\-cancel

NEWS
ModemManager 1.24.2
-------------------------------------------
This is a new stable release of ModemManager.

The most important features and changes in this release are the following:

 * Core:
   ** Wait until bearers are disconnected before allowing system to suspend
   ** Fix checks for CDMA/EVDO access technology in some modems
   ** Detect old Mediatek 62xx-based devices as GSM-capable
   ** Refresh registration info after manual registration for AT-based devices
   ** Correctly report NB-IOT access technology
   ** Disable unsolicited 3GPP events during quick suspend/resume

 * MBIM:
   ** Fix re-assembly of multi-part unstored SMS messages

 * QMI:
   ** Fix re-assembly of multi-part unstored SMS messages

 * Voice interface:
   ** Correctly start call state polling on some modems

 * Plugins:
   ** fibocomm: don't assert when checking for firmware interface
   ** cinterion: allow spaces in ^SXRAT response


ModemManager 1.24.0
-------------------------------------------
This is a new stable release of ModemManager.

The following notes are directed to package maintainers:

 * This version now requires:
   ** libmbim >= 1.32.0 (for the optional MBIM support)
   ** libqmi >= 1.36.0 (for the optional QMI support)

The most important features and changes in this release are the following:

 * API:
   ** Add new TIMEOUT, PROTOCOL, and THROTTLED core errors
   ** Add new MMCarrierLockError errors
   ** Add new Messaging interface SetDefaultStorage method
   ** Fix MM_BEARER_IP_FAMILY_ANY ABI break introduced in 1.20
   ** Add new Cell Broadcast API
   ** Report sub-system vendor and product IDs in Firmware device IDs
   ** Add new ID_MM_TTY_AT_PROBE_TRIES udev tag to control number of AT probes
      on a per-port basis for plugins that do not override generic probing logic.
   ** Add new Cell Broadcast API

 * Core:
   ** Improved handling of odd +CGDCONT responses
   ** Add support for putting modems in low-power mode during suspend
   ** Retry AT+CNUM a couple times if the SIM is busy
   ** Fix probing of modems that only expect a single AT port
   ** Reprobe modems on late port additions; ensures modems use the optimal
      control channel regardless of kernel/udev timing
   ** Ignore FASTBOOT ports during probing
   ** Better logging of user requests
   ** Fix probing of single-port devices by starting support check on first port
      addition
   ** Don't crash as often due to assumption that a primary modem port always
      exists
   ** Add support for loading and setting initial EPS bearer settings
   ** Quite a few memory leak fixes
   ** Lay groundwork for Intel XMM7xxx RPC-based devices
   ** restart AT probing if a port is stuck in PPP mode when modem is detected
   ** Serialize core modem operations to prevent multiple D-Bus clients from
      interrupting each other's operations
   ** Work around elogind sometimes not sending the PrepareForSleep signal
   ** Add generic Qualcomm Firehose and Sahara firmware update support
   ** Fix SIM unlock behavior being treated at SIM hot-swaps
   ** Detect 4G & 5G modem capabilities with newer AT commands

 * Translations:
   ** Updated Brazilian Portuguese and Polish translations

 * Modem interface:
   ** Logic to detect SIM swaps has been improved
   ** Do not set up carrier configs if the eSIM has no profiles
   ** Add delay between consecutive power state transitions for reliability
   ** Allow low-power state change while modem is enabled

 * 3GPP interface:
   ** Add 'NetworkRejection' property reporting errors during network
      registration

 * Location interface:
   ** Allow enabling the interface without a SIM

 * Profile Manager interface:
   ** Throttle updates from the modem to once every two seconds

 * Messaging interface:
   ** Validate supported storages
   ** Load default storage at initialization time

 * Bearer:
   ** Add new 'force' property to set EPS initial bearer settings

 * udev:
   ** Always ignore Qualcomm recovery layout devices (05c6:9008)
   ** Add ID_MM_QMI_PREALLOCATED_LINKS tag for NET ports
   ** Add ID_MM_QMI_PCO_DISABLED tag to disable PCO

 * mmcli:
   ** Higher long-running operation timeouts to avoid exiting before MM is done
   ** Print bearer D-Bus path after connection
   ** Add commands for getting/setting Cell Broadcast properties

 * libmm-glib:
   ** 3gpp-profile: fix loading roaming allowance and profile source

 * SMS:
   ** Ensure send/store always uses the same modem port
   ** Better re-assemble multi-part messages using both number and reference

 * Dispatcher:
   ** New 'setup' dispatcher runs after port probing but before ModemManager
      takes control

 * qrtr-bus-watcher:
   ** Increase wait time after probing to more reliably detect modems

 * MBIM:
   ** Improvements to SIM hot-swap detection and handling
   ** Fix re-opening MBIM ports if modem drops off the bus while being probed
   ** Don't die if ICCID is missing from the Subscriber Ready response
   ** Detect changes in carrier lock state
   ** Default to IPV4V6 dual-stack when IP type is unspecified
   ** Fix handling IP types from Provisioned Context Element V2 objects
   ** Don't ignore 'none' type profiles
   ** Use MBIMEx to list profiles if we can
   ** Detect SLAAC configuration updates to handle IPv6 connectivity on modems
      that take longer to run SLAAC themselves
   ** Add support for MS extensions device reset
   ** Update all attach settings (home/partner/non-partner) to the same values
      so that roaming attach uses same settings as home
   ** Read operator name from SIM EF files rather than from the network

cli/mmcli.c
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * mmcli -- Control modem status & access information from the command line
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (C) 2011-2023 Aleksander Morgado <aleksander@aleksander.es>
 * Copyright (C) 2011 Google, Inc.
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <locale.h>
#include <string.h>

#include <glib.h>
#include <gio/gio.h>

#define _LIBMM_INSIDE_MMCLI
#include <libmm-glib.h>

#include "mmcli.h"
#include "mmcli-common.h"
#include "mmcli-output.h"

#define PROGRAM_NAME    "mmcli"
#define PROGRAM_VERSION PACKAGE_VERSION

/* Globals */
static GMainLoop *loop;
static GCancellable *cancellable;

/* Main context */
static gboolean output_keyvalue_flag;
static gboolean output_json_flag;
static gboolean verbose_flag;
static gboolean version_flag;
static gboolean async_flag;
static gint timeout = 30; /* by default, use 30s for all operations */

static GOptionEntry main_entries[] = {
    { "output-keyvalue", 'K', 0, G_OPTION_ARG_NONE, &output_keyvalue_flag,
      "Run action with machine-friendly key-value output",
      NULL
    },
    { "output-json", 'J', 0, G_OPTION_ARG_NONE, &output_json_flag,
      "Run action with machine-friendly json output",
      NULL
    },
    { "verbose", 'v', 0, G_OPTION_ARG_NONE, &verbose_flag,
      "Run action with verbose logs",
      NULL
    },
    { "version", 'V', 0, G_OPTION_ARG_NONE, &version_flag,
      "Print version",
      NULL
    },
    { "async", 'a', 0, G_OPTION_ARG_NONE, &async_flag,
      "Use asynchronous methods",
      NULL
    },
    { "timeout", 0, 0, G_OPTION_ARG_INT, &timeout,
      "Timeout for the operation",
      "[SECONDS]"
    },
    { NULL }
};

/* Test context */
static gboolean test_session_flag;

static GOptionEntry test_entries[] = {
    { "test-session", 0, 0, G_OPTION_ARG_NONE, &test_session_flag,
      "Run in session DBus",
      NULL
    },
    { NULL }
};

static GOptionGroup *
test_get_option_group (void)
{
    GOptionGroup *group;

    group = g_option_group_new ("test",
                                "Test options:",
                                "Show test options",
                                NULL,
                                NULL);
    g_option_group_add_entries (group, test_entries);
    return group;
}

static void
signals_handler (int signum)
{
    if (cancellable) {
        /* Ignore consecutive requests of cancellation */
        if (!g_cancellable_is_cancelled (cancellable)) {
            g_printerr ("%s\n",
                        "cancelling the operation...\n");
            g_cancellable_cancel (cancellable);
        }
        return;
    }

    if (loop &&
        g_main_loop_is_running (loop)) {
        g_printerr ("%s\n",
                    "cancelling the main loop...\n");
        g_main_loop_quit (loop);
    }
}

static void
log_handler (const gchar *log_domain,
             GLogLevelFlags log_level,
             const gchar *message,
             gpointer user_data)
{
    const gchar *log_level_str;
    time_t now;
    gchar time_str[64];
    struct tm    *local_time;

    now = time ((time_t *) NULL);
    local_time = localtime (&now);
    strftime (time_str, 64, "%d %b %Y, %H:%M:%S", local_time);

    switch (log_level) {
    case G_LOG_LEVEL_WARNING:
        log_level_str = "-Warning **";
        break;

    case G_LOG_LEVEL_CRITICAL:
    case G_LOG_LEVEL_ERROR:
        log_level_str = "-Error **";
        break;

    case G_LOG_LEVEL_DEBUG:
        log_level_str = "[Debug]";
        break;

    case G_LOG_LEVEL_MESSAGE:
    case G_LOG_LEVEL_INFO:
        log_level_str = "";
        break;

    case G_LOG_FLAG_FATAL:
    case G_LOG_LEVEL_MASK:
    case G_LOG_FLAG_RECURSION:
    default:
        g_assert_not_reached ();
    }

    g_print ("[%s] %s %s\n", time_str, log_level_str, message);
}

static void
print_version_and_exit (void)
{
    g_print (PROGRAM_NAME " " PROGRAM_VERSION "\n"
             "Copyright (2011 - 2023) Aleksander Morgado\n"
             "License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl-2.0.html>\n"
             "This is free software: you are free to change and redistribute it.\n"
             "There is NO WARRANTY, to the extent permitted by law.\n"
             "\n");
    exit (EXIT_SUCCESS);
}

void
mmcli_async_operation_done (void)
{
    if (cancellable) {
        g_object_unref (cancellable);
        cancellable = NULL;
    }

    g_main_loop_quit (loop);
}

void
mmcli_force_async_operation (void)
{
    if (!async_flag) {
        g_debug ("Forcing request to be run asynchronously");
        async_flag = TRUE;
    }
}

void
mmcli_force_sync_operation (void)
{
    if (async_flag) {
        g_debug ("Ignoring request to run asynchronously");
        async_flag = FALSE;
    }
}

void
mmcli_force_operation_timeout (GDBusProxy *proxy)
{
    if (proxy)
        g_dbus_proxy_set_default_timeout (proxy, timeout * 1000);
}

gint
main (gint argc, gchar **argv)
{
    GDBusConnection *connection;
    GOptionContext *context;
    GError *error = NULL;

    setlocale (LC_ALL, "");

    /* Setup option context, process it and destroy it */
    context = g_option_context_new ("- Control and monitor the ModemManager");
    g_option_context_add_group (context,
                                mmcli_manager_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_get_common_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_3gpp_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_3gpp_profile_manager_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_3gpp_ussd_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_cdma_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_simple_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_location_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_messaging_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_voice_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_time_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_firmware_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_sar_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_signal_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_modem_oma_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_sim_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_bearer_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_sms_get_option_group ());
    g_option_context_add_group (context,
                                mmcli_call_get_option_group ());
    g_option_context_add_group (context,
                                test_get_option_group ());
    g_option_context_add_main_entries (context, main_entries, NULL);
    g_option_context_parse (context, &argc, &argv, NULL);
    g_option_context_free (context);

    if (version_flag)
        print_version_and_exit ();

    if (verbose_flag)
        g_log_set_handler (G_LOG_DOMAIN, G_LOG_LEVEL_MASK, log_handler, NULL);

    /* Setup output */
    if (output_keyvalue_flag && output_json_flag) {
        g_printerr ("error: only one output type supported at the same time\n");
        exit (EXIT_FAILURE);
    }
    if (output_keyvalue_flag) {
        if (verbose_flag) {
            g_printerr ("error: cannot set verbose output in keyvalue output type\n");
            exit (EXIT_FAILURE);
        }
        mmcli_output_set (MMC_OUTPUT_TYPE_KEYVALUE);
    }
    else if (output_json_flag) {
        if (verbose_flag) {
            g_printerr ("error: cannot set verbose output in JSON output type\n");
            exit (EXIT_FAILURE);
        }
        mmcli_output_set (MMC_OUTPUT_TYPE_JSON);
    } else {
        mmcli_output_set (MMC_OUTPUT_TYPE_HUMAN);
    }

    /* Setup signals */
    signal (SIGINT, signals_handler);
    signal (SIGHUP, signals_handler);
    signal (SIGTERM, signals_handler);

    /* Setup dbus connection to use */
    connection = g_bus_get_sync (test_session_flag ? G_BUS_TYPE_SESSION : G_BUS_TYPE_SYSTEM, NULL, &error);
    if (!connection) {
        g_printerr ("error: couldn't get bus: %s\n",
                    error ? error->message : "unknown error");
        exit (EXIT_FAILURE);
    }

    /* Create requirements for async options */
    cancellable = g_cancellable_new ();
    loop = g_main_loop_new (NULL, FALSE);

    /* Manager options? */
    if (mmcli_manager_options_enabled ()) {
        /* Ensure options from different groups are not enabled */
        if (mmcli_modem_options_enabled ()) {
            g_printerr ("error: cannot use manager and modem options "
                        "at the same time\n");
            exit (EXIT_FAILURE);
        }

        if (async_flag)
            mmcli_manager_run_asynchronous (connection, cancellable);
        else
            mmcli_manager_run_synchronous (connection);
    }
    /* Sim options? */
    else if (mmcli_sim_options_enabled ()) {
        if (async_flag)
            mmcli_sim_run_asynchronous (connection, cancellable);
        else
            mmcli_sim_run_synchronous (connection);
    }
    /* Bearer options? */
    else if (mmcli_bearer_options_enabled ()) {
        if (async_flag)
            mmcli_bearer_run_asynchronous (connection, cancellable);
        else
            mmcli_bearer_run_synchronous (connection);
    }
    /* Sms options? */
    else if (mmcli_sms_options_enabled ()) {
        if (async_flag)
            mmcli_sms_run_asynchronous (connection, cancellable);
        else
            mmcli_sms_run_synchronous (connection);
    }
    /* Call options? */
    else if (mmcli_call_options_enabled ()) {
        if (async_flag)
            mmcli_call_run_asynchronous (connection, cancellable);
        else
            mmcli_call_run_synchronous (connection);
    }
    /* Modem 3GPP options? */
    else if (mmcli_modem_3gpp_options_enabled ()) {
        if (async_flag)
            mmcli_modem_3gpp_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_3gpp_run_synchronous (connection);
    }
    /* Modem 3GPP profile manager options? */
    else if (mmcli_modem_3gpp_profile_manager_options_enabled ()) {
        if (async_flag)
            mmcli_modem_3gpp_profile_manager_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_3gpp_profile_manager_run_synchronous (connection);
    }
    /* Modem 3GPP USSD options? */
    else if (mmcli_modem_3gpp_ussd_options_enabled ()) {
        if (async_flag)
            mmcli_modem_3gpp_ussd_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_3gpp_ussd_run_synchronous (connection);
    }
    /* Modem CDMA options? */
    else if (mmcli_modem_cdma_options_enabled ()) {
        if (async_flag)
            mmcli_modem_cdma_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_cdma_run_synchronous (connection);
    }
    /* Modem Simple options? */
    else if (mmcli_modem_simple_options_enabled ()) {
        if (async_flag)
            mmcli_modem_simple_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_simple_run_synchronous (connection);
    }
    /* Modem Location options? */
    else if (mmcli_modem_location_options_enabled ()) {
        if (async_flag)
            mmcli_modem_location_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_location_run_synchronous (connection);
    }
    /* Modem Messaging options? */
    else if (mmcli_modem_messaging_options_enabled ()) {
        if (async_flag)
            mmcli_modem_messaging_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_messaging_run_synchronous (connection);
    }
    /* Voice options? */
    else if (mmcli_modem_voice_options_enabled ()) {
        if (async_flag)
            mmcli_modem_voice_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_voice_run_synchronous (connection);
    }
    /* Modem Time options? */
    else if (mmcli_modem_time_options_enabled ()) {
        if (async_flag)
            mmcli_modem_time_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_time_run_synchronous (connection);
    }
    /* Modem Firmware options? */
    else if (mmcli_modem_firmware_options_enabled ()) {
        if (async_flag)
            mmcli_modem_firmware_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_firmware_run_synchronous (connection);
    }
    /* Modem SAR options? */
    else if (mmcli_modem_sar_options_enabled ()) {
        if (async_flag)
            mmcli_modem_sar_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_sar_run_synchronous (connection);
    }
    /* Modem Signal options? */
    else if (mmcli_modem_signal_options_enabled ()) {
        if (async_flag)
            mmcli_modem_signal_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_signal_run_synchronous (connection);
    }
    /* Modem Oma options? */
    else if (mmcli_modem_oma_options_enabled ()) {
        if (async_flag)
            mmcli_modem_oma_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_oma_run_synchronous (connection);
    }
    /* Modem options?
     * NOTE: let this check be always the last one, as other groups also need
     * having a modem specified, and therefore if -m is set, modem options
     * are always enabled. */
    else if (mmcli_modem_options_enabled ()) {
        if (async_flag)
            mmcli_modem_run_asynchronous (connection, cancellable);
        else
            mmcli_modem_run_synchronous (connection);
    }
    /* No options? */
    else {
        g_printerr ("error: no actions specified\n");
        exit (EXIT_FAILURE);
    }

    /* Run loop only in async operations */
    if (async_flag)
        g_main_loop_run (loop);

    if (mmcli_manager_options_enabled ()) {
        mmcli_manager_shutdown ();
    } else if (mmcli_modem_3gpp_options_enabled ()) {
        mmcli_modem_3gpp_shutdown ();
    } else if (mmcli_modem_3gpp_profile_manager_options_enabled ()) {
        mmcli_modem_3gpp_profile_manager_shutdown ();
    } else if (mmcli_modem_3gpp_ussd_options_enabled ()) {
        mmcli_modem_3gpp_ussd_shutdown ();
    } else if (mmcli_modem_cdma_options_enabled ()) {
        mmcli_modem_cdma_shutdown ();
    } else if (mmcli_modem_simple_options_enabled ()) {
        mmcli_modem_simple_shutdown ();
    } else if (mmcli_modem_location_options_enabled ()) {
        mmcli_modem_location_shutdown ();
    } else if (mmcli_modem_messaging_options_enabled ()) {
        mmcli_modem_messaging_shutdown ();
    } else if (mmcli_modem_voice_options_enabled ()) {
        mmcli_modem_voice_shutdown ();
    } else if (mmcli_modem_time_options_enabled ()) {
        mmcli_modem_time_shutdown ();
    } else if (mmcli_modem_firmware_options_enabled ()) {
        mmcli_modem_firmware_shutdown ();
    } else if (mmcli_modem_sar_options_enabled ()) {
        mmcli_modem_sar_shutdown ();
    } else if (mmcli_modem_signal_options_enabled ()) {
        mmcli_modem_signal_shutdown ();
    } else if (mmcli_modem_oma_options_enabled ()) {
        mmcli_modem_oma_shutdown ();
    } else if (mmcli_sim_options_enabled ()) {
        mmcli_sim_shutdown ();
    } else if (mmcli_bearer_options_enabled ()) {
        mmcli_bearer_shutdown ();
    } else if (mmcli_sms_options_enabled ()) {
        mmcli_sms_shutdown ();
    } else if (mmcli_call_options_enabled ()) {
        mmcli_call_shutdown ();
    } else if (mmcli_modem_options_enabled ()) {
        mmcli_modem_shutdown ();
    }

    if (cancellable)
        g_object_unref (cancellable);
    g_main_loop_unref (loop);
    g_object_unref (connection);

    return EXIT_SUCCESS;
}
